python/ray/util/multiprocessing/pool.py

```diff
@@ -16,6 +26,102 @@
 RAY_ADDRESS_ENV = "RAY_ADDRESS"
 
 
+def _put_in_dict_registry(
+        obj: Any, registry_hashable: Dict[Hashable, Tuple]) -> ray.ObjectRef:
+    if obj not in registry_hashable:
+        ret = ray.put(obj)
+        registry_hashable[obj] = ret
+    else:
+        ret = registry_hashable[obj]
+    return ret
+
+
+def _put_in_list_registry(
+        obj: Any, registry: List[Tuple[Any, ray.ObjectRef]]) -> ray.ObjectRef:
+    try:
+        ret = next((ref for o, ref in registry if o is obj))
+    except StopIteration:
+        print(f"putting {obj} in store")
+        ret = ray.put(obj)
+        registry.append((obj, ret))
+    return ret
+
+
+def ray_put_if_needed(
+        obj: Any,
+        registry: Optional[List[Tuple[Any, ray.ObjectRef]]] = None,
+        registry_hashable: Optional[Dict[Hashable, Tuple]] = None
+) -> ray.ObjectRef:
+    """ray.put obj in object store if it's not an ObjRef and bigger than 100 bytes,
+    with support for list and dict registries"""
+    if isinstance(obj, ray.ObjectRef) or sys.getsizeof(obj) < 100:
+        return obj
+    ret = obj
+    if registry_hashable is not None:
+        try:
+            ret = _put_in_dict_registry(obj, registry_hashable)
+        except TypeError:
+            if registry is not None:
+                ret = _put_in_list_registry(obj, registry)
+    elif registry is not None:
+        ret = _put_in_list_registry(obj, registry)
+    return ret
+
+
+def ray_get_if_needed(obj: Any) -> Any:
+    """If obj is an ObjectRef, do ray.get, otherwise return obj"""
+    if isinstance(obj, ray.ObjectRef):
```

Comment:
Can we check here ObjectRef and ClientObjectRef?

Reply:
I think @mwtian has unified the types so we should not explicitly check for `ClientObjectRef` (This line of code fine as is.)
@mwtian to confirm.

Reply:
+1 on what Dmitri said, this should be fine.

Reply:
I agree, but IIRC, recently we had this bug where it failed for that reason, but I guess it was the wrong ray version?

can you also pls review the rest of the PR?